web/src/service-worker/index.ts

```diff
@@ -0,0 +1,137 @@
+  event.waitUntil(addFilesToCache());
+});
+
+sw.addEventListener('activate', (event) => {
+  event.waitUntil(sw.clients.claim());
+  // Remove previous cached data from disk
+  event.waitUntil(deleteOldCaches());
+});
+
+sw.addEventListener('fetch', (event) => {
+  // ignore POST requests etc
+  if (event.request.method !== 'GET') return;
+  const url = new URL(event.request.url);
+  if (APP_RESOURCES.includes(url.pathname)) {
+    event.respondWith(appResources(url, event));
+  } else if (/^\/api\/assets\/[a-f0-9-]+\/(original|thumbnail)/.test(url.pathname)) {
+    event.respondWith(immichAsset(url));
+  }
+});
+
+async function deleteOldCaches() {
+  for (const key of await caches.keys()) {
+    if (key !== CACHE) await caches.delete(key);
+  }
+}
+
+async function addFilesToCache() {
+  const cache = await caches.open(CACHE);
+  await cache.addAll(APP_RESOURCES);
+}
+
+async function immichAsset(url: URL) {
+  const cache = await caches.open(CACHE);
+  let response = useCache ? await cache.match(url) : undefined;
+  if (!response) {
+    try {
+      const cancelToken = new AbortController();
+      const request = fetch(url, {
+        signal: cancelToken.signal,
+      });
+      pendingLoads.set(url.toString(), cancelToken);
+      response = await request;
+      if (!(response instanceof Response)) {
+        throw new TypeError('invalid response from fetch');
+      }
+      if (response.status === 200) {
+        cache.put(url, response.clone());
+      }
+    } catch (error) {
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      if ((error as any).name !== 'AbortError') {
+        throw error;
+      }
+    } finally {
+      pendingLoads.delete(url.toString());
+    }
+  }
+  return response as Response;
+}
+
+async function appResources(url: URL, event: FetchEvent) {
+  const cache = await caches.open(CACHE);
+  // `build`/`files` can always be served from the cache
+  if (APP_RESOURCES.includes(url.pathname)) {
+    const response = await cache.match(url.pathname);
+    if (response) {
+      return response;
+    }
+  }
+  // for everything else, try the network first, but
+  // fall back to the cache if we're offline
+  try {
+    const response = await fetch(event.request);
+    // if we're offline, fetch can return a value that is not a Response
+    // instead of throwing - and we can't pass this non-Response to respondWith
+    if (!(response instanceof Response)) {
+      throw new TypeError('invalid response from fetch');
+    }
+
+    if (response.status === 200) {
+      cache.put(event.request, response.clone());
+    }
+
+    return response;
+  } catch (error) {
+    const response = await cache.match(event.request);
+    if (response) {
+      return response;
+    }
+    // if there's no cache, then just error out
+    // as there is nothing we can do to respond to this request
+    throw error;
+  }
+}
+
+const broadcast = new BroadcastChannel('immich');
```

Comment:
Is this interfere with `web/src/lib/utils/sw-messaging.ts`?

Reply:
Yes. This is really simple for now. I foresee this growing in future, and can be improved as it grows.

Reply:
Oh, i misread your comment - I read *interface* instead of interfere. No - BroadcastChannel is a pub/sub api. the name of the channel is effectively the topic. Anyone that creates a broadcast channel with the same name can listen or send messages to it. There is no need to set up the sender/receiver in advance. i.e. a sender can send a message to 0:m receivers.